appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockLineChartDataModel.java

```diff
@@ -0,0 +1,108 @@
+     * @param chartData  Data object of the Chart View.
+     */
+    public MockLineChartDataModel(Data chartData) {
+        super(chartData);
+
+        // Create the Data Series object
+        dataSeries = new ScatterDataset();
+
+        // Set the default style properties for the Data Series
+        setDefaultStylingProperties();
+
+        // Adds the Data Series to the Chart.
+        addDataSeriesToChart();
+    }
+
+    @Override
+    public void changeColor(String color) {
+        color = getHexColor(color);
+        dataSeries.setBackgroundColor(color);
+        dataSeries.setPointBackgroundColor(color);
+        dataSeries.setBorderColor(color);
+    }
+
+    @Override
+    public void setElements(String elements) {
+        String[] entries = elements.split(",");
+
+        // Create new list of Data Points
+        ArrayList<DataPoint> dataPoints = new ArrayList<DataPoint>();
+
+        // Since entries come in pairs, we add 2 on each iteration.
+        // Beginning from i = 1 instead of 0 to prevent out of bounds
+        // accesses.
+        for (int i = 1; i < entries.length; i += 2) {
+            try {
+                DataPoint dataPoint = new DataPoint();
+                dataPoint.setX(Double.parseDouble(entries[i-1]));
+                dataPoint.setY(Double.parseDouble(entries[i]));
+                dataPoints.add(dataPoint);
+            } catch (NumberFormatException e) {
+                return; // Wrong input. Do not update entries.
+            }
+        }
+
+        // No data points generated, fallback to default option.
+        if (dataPoints.isEmpty()) {
+            setDefaultElements(dataPoints);
+        } else {
+            // Since we are dealing with a Scatter Data Series, sorting
+            // is a must, because otherwise, the Chart will not look representative.
+            // Consider adding: (1, 2), (5, 3), (2, 5). We want the x = 2
+            // value to be continuous on the Line Chart, rather than
+            // going outside the Chart, which would happen since we
+            // are using a Scatter Chart.
+            dataPoints.sort(Comparator.comparingDouble(DataPoint::getX));
+        }
+
+        // Set the generated data points to the Data Series
+        dataSeries.setDataPoints(dataPoints);
+    }
+
+    @Override
+    protected void setDefaultElements(List<DataPoint> dataPoints) {
+        final int points = 4; // Number of points to add
+
+        // TBD: Might change this in the future.
+        // Generally, this should not cause performance issues because typically there are not
+        // that many data points.
+        Optional<DataPoint> maxYPoint = chartData.getDatasets() // Get all the data sets
+                .stream() // Create a stream
+                .flatMap(l -> ((ScatterDataset)l).getDataPoints().stream()) // Flatten the nested lists to a List of data points
+                .max(Comparator.comparing(DataPoint::getY)); // Get the maximum data point value
```

Comment:
Two questions:
1. Is it fine to use lambda expressions and streams here? I found it does the job quickly.
2. Is this too inefficient of an approach? This takes all the data points in the Data Set, so given large Data Sets, it might be too cumbersome. However, large data sets are not very realistic in the Designer, given most Data comes from external sources that only get triggered in Android itself.

Reply:
An alternative is to keep track of the largest value (although that requires logic in more than few places), but I suppose that would be the better approach here?